File: app/src/main/java/io/nekohasekai/sagernet/fmt/wireguard/WireGuardBean.java
package io.nekohasekai.sagernet.fmt.wireguard;

import androidx.annotation.NonNull;

import com.esotericsoftware.kryo.io.ByteBufferInput;
import com.esotericsoftware.kryo.io.ByteBufferOutput;

import io.nekohasekai.sagernet.fmt.AbstractBean;
import io.nekohasekai.sagernet.fmt.KryoConverters;
import moe.matsuri.nb4a.SingBoxOptions;

import org.jetbrains.annotations.NotNull;

public class WireGuardBean extends AbstractBean {

    public static final Creator<WireGuardBean> CREATOR = new CREATOR<WireGuardBean>() {
        @NonNull
        @Override
        public WireGuardBean newInstance() {
            return new WireGuardBean();
        }

        @Override
        public WireGuardBean[] newArray(int size) {
            return new WireGuardBean[size];
        }
    };
    public String localAddress;
    public String privateKey;
    public String publicKey;
    public String preSharedKey;
    public Integer mtu;
    public String reserved;
    /**
     * Enable listen if it > 0
     */
    public Integer listenPort;
    public Integer persistentKeepaliveInterval;

    @Override
    public void initializeDefaultValues() {
        super.initializeDefaultValues();
        if (localAddress == null) localAddress = "";
        if (privateKey == null) privateKey = "";
        if (publicKey == null) publicKey = "";
        if (preSharedKey == null) preSharedKey = "";
        if (mtu == null || mtu < 1000 || mtu > 65535) mtu = 1420;
        if (reserved == null) reserved = "";
        if (listenPort == null) listenPort = 0;
        if (persistentKeepaliveInterval == null) persistentKeepaliveInterval = 0;
    }

    @Override
    public void serialize(ByteBufferOutput output) {
        output.writeInt(2);
        super.serialize(output);
        output.writeString(localAddress);
        output.writeString(privateKey);
        output.writeString(publicKey);
        output.writeString(preSharedKey);
        output.writeInt(mtu);
        output.writeString(reserved);
        output.writeInt(listenPort);
        output.writeInt(persistentKeepaliveInterval);
    }

    @Override
    public void deserialize(ByteBufferInput input) {
        int version = input.readInt();
        super.deserialize(input);
        localAddress = input.readString();
        privateKey = input.readString();
        publicKey = input.readString();
        preSharedKey = input.readString();
        mtu = input.readInt();
        reserved = input.readString();
        if (version >= 1) {
            listenPort = input.readInt();
        }
        if (version >= 2) {
            persistentKeepaliveInterval = input.readInt();
        }
    }

    @Override
    public boolean canTCPing() {
        return false;
    }

    @NotNull
    @Override
    public WireGuardBean clone() {
        return KryoConverters.deserialize(new WireGuardBean(), KryoConverters.serialize(this));
    }

    @Override
    public @NotNull String outboundType() throws Throwable {
        return SingBoxOptions.TYPE_WIREGUARD;
    }
}

File: app/src/main/kotlin/io/nekohasekai/sagernet/fmt/wireguard/WireGuardBean.kt
package io.nekohasekai.sagernet.fmt.wireguard

import com.esotericsoftware.kryo.io.ByteBufferInput
import com.esotericsoftware.kryo.io.ByteBufferOutput
import io.nekohasekai.sagernet.fmt.AbstractBean
import io.nekohasekai.sagernet.fmt.KryoConverters

class WireGuardBean : AbstractBean() {

    companion object {
        @JvmField
        val CREATOR = object : CREATOR<WireGuardBean>() {
            override fun newInstance(): WireGuardBean {
                return WireGuardBean()
            }

            override fun newArray(size: Int): Array<WireGuardBean?> {
                return arrayOfNulls(size)
            }
        }
    }

    var localAddress: String = ""
    var privateKey: String = ""
    var publicKey: String = ""
    var preSharedKey: String = ""
    var mtu: Int = 1420
    var reserved: String = ""

    /**
     * Enable listen if it > 0
     */
    var listenPort: Int = 0
    var persistentKeepaliveInterval: Int = 0

    override fun initializeDefaultValues() {
        super.initializeDefaultValues()
        if (mtu !in 1000..65535) mtu = 1420
    }

    override fun serialize(output: ByteBufferOutput) {
        output.writeInt(2)
        super.serialize(output)
        output.writeString(localAddress)
        output.writeString(privateKey)
        output.writeString(publicKey)
        output.writeString(preSharedKey)
        output.writeInt(mtu)
        output.writeString(reserved)
        output.writeInt(listenPort)
        output.writeInt(persistentKeepaliveInterval)
    }

    override fun deserialize(input: ByteBufferInput) {
        val version = input.readInt()
        super.deserialize(input)
        localAddress = input.readString()
        privateKey = input.readString()
        publicKey = input.readString()
        preSharedKey = input.readString()
        mtu = input.readInt()
        reserved = input.readString()
        if (version >= 1) {
            listenPort = input.readInt()
        }
        if (version >= 2) {
            persistentKeepaliveInterval = input.readInt()
        }
    }

    override val canTCPing = false

    override fun clone(): WireGuardBean {
        return KryoConverters.deserialize(WireGuardBean(), KryoConverters.serialize(this))
    }

    override val defaultPort = 51820
}
